File: backend/src/agents/base-agent.ts
import { Logger } from '../utils/logger';
import { db, Collections } from '../services/firestore';
import { eventFabric } from '../services/event-fabric';
import crypto from 'crypto';

export enum AgentState {
  IDLE = 'idle',
  THINKING = 'thinking',
  WORKING = 'working',
  DONE = 'done',
  ERROR = 'error'
}

export interface AgentConfig {
  id: string;
  name: string;
  role?: string;
  model?: string;
  color?: string;
  systemInstruction?: string;
}

export type AgentOutputType = 
  | "copy" | "image_prompt" | "strategy" 
  | "audit" | "video" | "analysis" | "blog_post";

export abstract class BaseAgent {
  protected config: AgentConfig;
  protected logger: Logger;
  protected state: AgentState = AgentState.IDLE;
  protected progress: number = 0;
  protected lastStatus: string = '';
  public onStatusUpdate?: (status: any) => void;

  // Operational context
  protected runId: string | null = null;
  protected campaignId: string | null = null;

  constructor(config: AgentConfig | string, name?: string, color?: string) {
    if (typeof config === 'string') {
      this.config = { id: config, name: name || config, color };
    } else {
      this.config = config;
    }
    this.logger = new Logger(this.config.id);
  }

  /**
   * Sets the operational context for the agent.
   */
  public setContext(runId: string, campaignId?: string) {
    this.runId = runId;
    this.campaignId = campaignId || null;
    this.logger.info(`Context established: Run=${runId} Campaign=${campaignId || 'none'}`);
  }

  /**
   * Core execution logic for the agent.
   * Every agent in the GenIUS swarm must implement this.
   */
  abstract execute(input: any): Promise<any>;

  /**
   * Status reporting for the React Console and Firestore.
   */
  protected async updateStatus(state: AgentState, status: string, progress: number = 0) {
    this.state = state;
    this.lastStatus = status;
    this.progress = progress;
    this.logger.info(`[${state.toUpperCase()}] ${status} (${progress}%)`);
    
    const currentStatus = this.getStatus();
    
    // 1. Broadcast to WebSockets
    eventFabric.broadcastStatus(currentStatus);

    // 2. Persist to Firestore /agent_status/{id}
    try {
      await db.collection(Collections.AGENT_STATUS).doc(this.id).set({
        agent_id: this.id,
        status: state,
        message: status,
        progress: progress,
        run_id: this.runId,
        updated_at: new Date()
      }, { merge: true });
    } catch (e) {
      this.logger.error(`Failed to persist status to Firestore: ${e}`);
    }

    if (this.onStatusUpdate) {
      this.onStatusUpdate(currentStatus);
    }
  }

  /**
   * Writes the final output of an agent to Firestore /agent_outputs.
   * This is the single source of truth for all module updates.
   */
  protected async writeOutput(
    type: AgentOutputType,
    payload: any,
    taskId?: string,
    senateRequired: boolean = false
  ): Promise<string> {
    const outputId = crypto.randomUUID();
    const now = new Date();

    const outputDoc = {
      output_id: outputId,
      run_id: this.runId,
      campaign_id: this.campaignId,
      task_id: taskId ?? null,
      agent_id: this.id,
      agent_name: this.config.name,
      type: type,
      payload: payload,
      created_at: now,
      senate_status: senateRequired ? 'pending' : 'approved',
      visible_in_ui: !senateRequired
    };

    this.logger.info(`Writing output [${type}] with ID: ${outputId}`);

    // 1. Save main output
    await db.collection(Collections.AGENT_OUTPUTS).doc(outputId).set(outputDoc);

    // 2. Update Swarm Run state if task_id exists
    if (taskId && this.runId) {
      try {
        const runRef = db.collection(Collections.SWARM_RUNS).doc(this.runId);
        await runRef.update({
          [`tasks.${taskId}.status`]: 'completed',
          [`tasks.${taskId}.output_id`]: outputId,
          [`tasks.${taskId}.completed_at`]: now
        });
      } catch (e) {
        this.logger.warn(`Could not update swarm run task: ${e}`);
      }
    }

    // 3. Add to Senate Queue if required
    if (senateRequired) {
      await db.collection(Collections.SENATE_QUEUE).doc(outputId).set({
        ...outputDoc,
        review_required: true
      });
    }

    // 4. Update Campaign Timeline
    if (this.campaignId) {
       try {
         // Using dynamic import of firebase-admin to stay consistent with other services
         const admin = require('firebase-admin');
         await db.collection(Collections.CAMPAIGNS).doc(this.campaignId).update({
           timeline: admin.firestore.FieldValue.arrayUnion({
             output_id: outputId,
             type: type,
             agent_id: this.id,
             created_at: now
           })
         });
       } catch (e) {
         this.logger.warn(`Could not update campaign timeline: ${e}`);
       }
    }

    // 5. Broadcast Event for immediate UI responsiveness
    eventFabric.broadcast({
      type: 'agent_output',
      data: {
        output_id: outputId,
        type: type,
        agent_id: this.id,
        run_id: this.runId,
        campaign_id: this.campaignId
      }
    });

    return outputId;
  }

  public get id(): string {
    return this.config.id;
  }

  public getStatus() {
    return {
      id: this.config.id,
      name: this.config.name,
      color: this.config.color,
      state: this.state,
      lastStatus: this.lastStatus,
      progress: this.progress
    };
  }

  public getConfig(): AgentConfig {
    return this.config;
  }
}

File: engine/agents/base_agent.py
from google.cloud import firestore
from datetime import datetime, timezone
import uuid
from typing import Any, Literal, Optional
import os

# Initialize Firestore Client
db = firestore.Client()

OutputType = Literal["copy", "image_prompt", "strategy", "audit", "video", "analysis", "blog_post"]

class BaseAgent:
    """
    Alle Agenten erben von hier.
    Die write_output-Methode ist die einzige Wahrheitsquelle für Output-Routing.
    """
    agent_id: str = "unknown"
    agent_name: str = "Unknown Agent"

    def __init__(self, run_id: str, campaign_id: Optional[str] = None):
        self.run_id = run_id
        self.campaign_id = campaign_id
        self.db = db

    async def write_output(
        self,
        output_type: OutputType,
        payload: dict[str, Any],
        task_id: Optional[str] = None,
        senate_required: bool = False,
    ) -> str:
        """
        Schreibt Agent-Output nach Firestore /agent_outputs/{id}.
        Gibt die output_id zurück.
        Wenn senate_required=True, wird zusätzlich /senate_queue/{id} geschrieben.
        """
        output_id = str(uuid.uuid4())
        now = datetime.now(timezone.utc)
        
        output_doc = {
            "output_id": output_id,
            "run_id": self.run_id,
            "campaign_id": self.campaign_id,
            "task_id": task_id,
            "agent_id": self.agent_id,
            "agent_name": self.agent_name,
            "type": output_type,
            "payload": payload,
            "created_at": now,
            "senate_status": "pending" if senate_required else "approved",
            "visible_in_ui": not senate_required,  # UI zeigt erst nach Senate-Approval
        }

        # Primäres Output-Dokument
        await self.db.collection("agent_outputs").document(output_id).set(output_doc)

        # Swarm Run Task-Status aktualisieren
        if task_id and self.run_id:
            run_ref = self.db.collection("swarm_runs").document(self.run_id)
            # Use dot notation for nested field update in Firestore
            await run_ref.update({
                f"tasks.{task_id}.status": "completed",
                f"tasks.{task_id}.output_id": output_id,
                f"tasks.{task_id}.completed_at": now,
            })

        # Senate Queue eintragen wenn nötig
        if senate_required:
            await self.db.collection("senate_queue").document(output_id).set({
                **output_doc,
                "senate_status": "pending",
                "review_required": True,
            })

        # Campaign Timeline updaten
        if self.campaign_id:
            campaign_ref = self.db.collection("campaigns").document(self.campaign_id)
            await campaign_ref.update({
                "timeline": firestore.ArrayUnion([{
                    "output_id": output_id,
                    "type": output_type,
                    "agent_id": self.agent_id,
                    "created_at": now,
                }])
            })

        # WebSocket Event feuern (für Echtzeit-UI-Updates ohne Page Reload)
        await self._emit_websocket_event("agent_output", {
            "output_id": output_id,
            "type": output_type,
            "agent_id": self.agent_id,
            "run_id": self.run_id,
            "campaign_id": self.campaign_id,
        })

        return output_id

    async def _emit_websocket_event(self, event_type: str, data: dict):
        """Sendet Event an alle verbundenen WebSocket-Clients."""
        # Note: In a real distributed system, this might go via Pub/Sub or a shared Redis
        # For now, we attempt to use the local ws_manager if in the same process, 
        # or we could rely on Firestore onSnapshot on the frontend.
        # But the Handbook suggests a broadcast.
        try:
            from core.ws_manager import ws_manager
            await ws_manager.broadcast({
                "event": event_type,
                "data": data,
                "timestamp": datetime.now(timezone.utc).isoformat(),
            })
        except ImportError:
            # If core.ws_manager is not available (e.g. running in a different container),
            # the frontend will still catch it via Firestore onSnapshot.
            pass

    async def update_status(self, status: str, meta: dict = {}):
        """Aktualisiert den Agent-Status in der Synergy Map."""
        await self.db.collection("agent_status").document(self.agent_id).set({
            "status": status,
            "agent_id": self.agent_id,
            "run_id": self.run_id,
            "updated_at": datetime.now(timezone.utc),
            **meta,
        }, merge=True)
        
        await self._emit_websocket_event("agent_status", {
            "agent_id": self.agent_id,
            "status": status,
        })
